### DISCUSSION TITLE
Invalidate In-Flight Server Actions using NextJs Middleware

### DISCUSSION THREAD

**Discussion by boncz92** [vercel/next.js#65148] [App Router]
As part of migrating to the app router and embracing the Server Component/Server Action features I wanted to move away from statically built Rest endpoint where I could and instead utilize Server Actions. In doing so I quickly realized that there is not functionality to intercept an in-flight Server action and redirect back to the client. In doing so, the client receives the payload, but not in a JSON format rather than the RSC format it expects, which makes the returned values unreachable.

As shown below:
![image](https://github.com/vercel/next.js/assets/8141438/5fe89e92-bf2e-4941-a083-69cb1cf060af)

A work around for this that is functiona is below:

Middleware:
```js
if (!authenticationCookie) {
        if (serverAction != null) {
            const response = NextResponse.next();
            response.headers.set('Middleware-Authentication', 'false')
            return response;
        }
}
```
Server Action Utility:
```js
import { headers } from 'next/headers';
 
 
export function serverActionAuth() {
    const authValue = headers().get('Middleware-Authentication');
    if (authValue == 'false') {
        return { Authentication: false };
    }
    else {
        return;
    }
}
```

Code to include in all Server Actions you want protected
```js
const auth = serverActionAuth()
if (auth) return auth;
```

I beleive that this solution adds a level complexity that shouldnt be required given that the middleware works as a perfect authentication layer for api calls and routing. I'd like to propose a change or helper function that would allow the middleware to invalidate a server action in-flight by adding a header to the request similarly to how I implemented it. 

Wanted to see if there were others having this same issue and think this is a viable solution.

[Relevant NextJs Source Code](https://github.com/vercel/next.js/blob/ae524fb24499fb8caf5b70eb8ce4cc96a5301565/packages/next/src/client/components/router-reducer/reducers/server-action-reducer.ts#L54)

**Comment by icyJoseph** [comment]
> I beleive that this solution adds a level complexity that shouldnt be required given that the middleware works as a perfect authentication layer for api calls and routing.

I think there was a lot of discussion regarding that belief... here: https://github.com/vercel/next.js/discussions/63775#discussioncomment-8932242

In short, assuming that a decoupled check, is enough to protect some data layer access, is bound to cause issues. You should, collocate verification and access. 

Check this post out: https://nextjs.org/blog/security-nextjs-server-components-actions the Write data section at the end is what you want, https://nextjs.org/blog/security-nextjs-server-components-actions#write ~
